portal-impl/src/com/liferay/portal/jsonwebservice/JSONWebServiceActionImpl.java

```diff
@@ -598,6 +609,96 @@ private Object _invokeActionMethod() throws Exception {
 		_jsonWebServiceActionParameters;
 	private final JSONWebServiceNaming _jsonWebServiceNaming;
 
+	private static class DateTypeConverter implements TypeConverter<Date> {
+
+		@Override
+		public Date convert(Object object) {
+			if (object == null) {
+				return null;
+			}
+
+			if (object instanceof Date) {
+				return (Date)object;
+			}
+
+			if (object instanceof Calendar) {
+				Calendar calendar = (Calendar)object;
+
+				return new Date(calendar.getTimeInMillis());
+			}
+
+			if (object instanceof LocalDateTime) {
+				return TimeUtil.toDate((LocalDateTime)object);
+			}
+
+			if (object instanceof LocalDate) {
+				return TimeUtil.toDate((LocalDate)object);
+			}
+
+			if (object instanceof Number) {
+				Number number = (Number)object;
+
+				return new Date(number.longValue());
+			}
+
+			String stringValue = object.toString();
+
+			stringValue = stringValue.trim();
+
+			if (!StringUtil.containsOnlyDigits(stringValue)) {
+				TemporalAccessor temporalAccessor =
+					_dateTimeFormatter.parseBest(
+						stringValue, ZonedDateTime::from, LocalDateTime::from,
+						LocalDate::from);
+
+				if (temporalAccessor instanceof LocalDate) {
+					return TimeUtil.toDate((LocalDate)temporalAccessor);
+				}
+
+				if (temporalAccessor instanceof LocalDateTime) {
+					return TimeUtil.toDate((LocalDateTime)temporalAccessor);
+				}
+
+				if (temporalAccessor instanceof ZonedDateTime) {
+					ZonedDateTime zonedDateTime =
+						(ZonedDateTime)temporalAccessor;
+
+					return Date.from(zonedDateTime.toInstant());
+				}
+
+				throw new TypeConversionException(object);
+			}
+
+			try {
+				long milliseconds = Long.parseLong(stringValue);
+
+				return new Date(milliseconds);
+			}
+			catch (NumberFormatException numberFormatException) {
+				throw new TypeConversionException(
+					object, numberFormatException);
+			}
+		}
+
+		private static final DateTimeFormatter _dateTimeFormatter =
```

Comment:
@dantewang Is it possible to use the predefined DateTimeFormatter? like DateTimeFormatter#ISO_LOCAL_DATE_TIME instead of creating one yourself?

Reply:
@tinatian No. Predefined formatters strictly parses the input. For example, the date part in `ISO_LOCAL_DATE_TIME` are defined as `ISO_LOCAL_DATE` which is defined by
```
        ISO_LOCAL_DATE = new DateTimeFormatterBuilder()
                .appendValue(YEAR, 4, 10, SignStyle.EXCEEDS_PAD)
                .appendLiteral('-')
                .appendValue(MONTH_OF_YEAR, 2)
                .appendLiteral('-')
                .appendValue(DAY_OF_MONTH, 2)
                .toFormatter(ResolverStyle.STRICT, IsoChronology.INSTANCE);
```
Here both `MONTH_OF_YEAR` and `DAY_OF_MONTH` was given width 2. So when parsing it will always try to get 2 digits, which will break on date like `2021-5-13`.

There's no pre-defined formatter to support date like `yyyy-M-d`.